projects/RabbitMQ.Client/client/api/ICredentialsRefresher.cs

```diff
@@ -80,65 +84,123 @@ public ICredentialsProvider Register(ICredentialsProvider provider, ICredentials
                 return provider;
             }
 
-            if (_registrations.TryAdd(provider, scheduleTimer(provider, callback)))
+            lock (_lockObj)
             {
+                if (_registrations.TryGetValue(provider, out var registration))
+                {
+                    registration.Callback = callback;
+                    TimerBasedCredentialRefresherEventSource.Log.AlreadyRegistered(provider.Name);
+                    return provider;
+                }
+
+                registration = new TimerRegistration(_lockObj, callback);
+                _registrations.Add(provider, registration);
+                registration.ScheduleTimer(provider);
+
                 TimerBasedCredentialRefresherEventSource.Log.Registered(provider.Name);
             }
-            else
-            {
-                TimerBasedCredentialRefresherEventSource.Log.AlreadyRegistered(provider.Name);
-            }
 
             return provider;
         }
 
         public bool Unregister(ICredentialsProvider provider)
         {
-            if (_registrations.TryRemove(provider, out System.Timers.Timer? timer))
+            lock (_lockObj)
             {
-                try
+                if (_registrations.TryGetValue(provider, out var registration))
                 {
+                    _registrations.Remove(provider);
+
                     TimerBasedCredentialRefresherEventSource.Log.Unregistered(provider.Name);
-                    timer.Stop();
-                }
-                finally
-                {
-                    timer.Dispose();
+                    registration.Dispose();
+                    return true;
                 }
-                return true;
-            }
-            else
-            {
-                return false;
             }
+
+            return false;
         }
 
-        private System.Timers.Timer scheduleTimer(ICredentialsProvider provider, ICredentialsRefresher.NotifyCredentialRefreshedAsync callback)
+        private class TimerRegistration : IDisposable
         {
-            System.Timers.Timer timer = new System.Timers.Timer();
-            timer.Interval = provider.ValidUntil!.Value.TotalMilliseconds * (1.0 - (1 / 3.0));
-            timer.Elapsed += (o, e) =>
+
+            private readonly object _lockObj;
+            private System.Timers.Timer? _timer;
+            private bool _disposed;
+
+            public ICredentialsRefresher.NotifyCredentialRefreshedAsync Callback { get; set; }
+
+            public TimerRegistration(object lockObj, ICredentialsRefresher.NotifyCredentialRefreshedAsync callback)
```

Comment:
This is actually something I noticed as well, but I'm quite unsure on what would be an alternative. Unregistering may happen at any time, even when the timer elapsed event is being triggered in parallel. When removing the lock the worst case would be, that the timer runs one more cycle and then shuts itself down. I guess this is acceptable in order to avoid passing the lock object down. Any other ideas?